### PR TITLE
oAuth: Added ability to augment the state parameter in oAuth requests

### LINKED ISSUE #32887: oAuth: ability to influence the state parameter
**What would you like to be added**:
I'd like to be able to influence the state parameter that is send with the oAuth requests

**Why is this needed**:
In a setup where there are multiple Grafana instances running and exposed via different urls (for example multiple K8S clusters, each with their own Grafana) and you'd like to use oAuth for authentication, without the ability to influence the state parameter, you'd either have to register the (callback) url for each Grafana instance individually on the oAuth client at the OP or have 1 oAuth client per Grafana instance at the OP

If you can add some identifier to the state parameter in the oAuth requests, you can have one generic redirect url, which then redirects the callback from the OP to the correct Grafana instance, based on the identifier in the state parameter

### PR DESCRIPTION
**Pull request by p-bakker**
**What this PR does / why we need it**:
See description in https://github.com/grafana/grafana/issues/32887

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/grafana/issues/32887

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/api/login_oauth.go
Comment (aknuds1) [inline pkg/api/login_oauth.go:84]: Use non-global config:

```suggestion
		stateConfig := hs.Cfg.OAuthService.OAuthInfos[name].State
```

**Review by aknuds1 — DISMISSED** [review]

File: docs/sources/auth/generic-oauth.md
Comment (achatterjee-grafana) [inline docs/sources/auth/generic-oauth.md:50]:
```suggestion
Grafana generates a unique `state` value, for security purposes. You can also modify the value of the state parameter.
- if `state` includes the text `<random>`, the text `<random>` is replaced with a randomly generated value.
- if `state` does not include the text `<random>`, an automatically generated value is appended to the `state` value.
- if no value is provided, `state` is set to automatically generated random value.

```

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Added doc change request.

**Comment by sakjur** [comment]
@aknuds1 For CSP, `$NONCE` was introduced as the magic string for the random value. Would it make sense to `s/<random>/$NONCE` here to stay internally consistent? https://github.com/grafana/grafana/blob/258578766b43edd042c3d6e6ad4154ad796fe244/pkg/middleware/csp.go#L45

**Comment by p-bakker** [comment]
In oAuth a nonce is serving a different purpose as a state parameter, so imho it wouldn't make sense to use the `nonce` here. 

If in other places there's already some dynamic string replacement going on using $xxxx, then I think the same should be used here, so `$random` vs. `<random>`

File: pkg/api/login_oauth.go
Comment (p-bakker) [reply pkg/api/login_oauth.go]: done

File: docs/sources/auth/generic-oauth.md
Comment (p-bakker) [reply docs/sources/auth/generic-oauth.md]: done

**Comment by aknuds1** [comment]
PR still doesn't build.

**Comment by p-bakker** [comment]
I fixed one problem why the pipeline was failing, but there's another problem: the non-global config object `hs.Cfg.` doesn't have `OAuthService`...

So, do you expect me to add that within the scope of this PR and refactor all existing code that accesses `OAuthService` via the global config?

**Comment by aknuds1** [comment]
> I fixed one problem why the pipeline was failing, but there's another problem: the non-global config object `hs.Cfg.` doesn't have `OAuthService`...
> 
> So, do you expect me to add that within the scope of this PR and refactor all existing code that accesses `OAuthService` via the global config?

Hang on until we've decided if we want the functionality or not. If we do, I think that setting should be added to the `setting.Cfg` type. No need to refactor all other code though.

**Comment by p-bakker** [comment]
@aknuds1 ok. Any idea when that decision might be made?

File: docs/sources/auth/generic-oauth.md
Comment (oddlittlebird) [reply docs/sources/auth/generic-oauth.md]: Please check this doc with a local build. I had an issue the other day where text enclosed in pointy braces did not appear when the page was generated in Hugo.

File: docs/sources/auth/generic-oauth.md
Comment (oddlittlebird) [inline docs/sources/auth/generic-oauth.md:47]:
```suggestion
- If `state` includes the text `<random>`, then the text `<random>` is replaced with a randomly-generated value.
```

File: docs/sources/auth/generic-oauth.md
Comment (oddlittlebird) [inline docs/sources/auth/generic-oauth.md:48]:
```suggestion
- If `state` does not include the text `<random>`, then an automatically-generated value is appended to the `state` value.
```

File: docs/sources/auth/generic-oauth.md
Comment (oddlittlebird) [inline docs/sources/auth/generic-oauth.md:49]:
```suggestion
- If no value is provided, then `state` is set to an automatically-generated random value.
```

**Review by oddlittlebird — COMMENTED** [review]
Minor format edits

File: docs/sources/auth/generic-oauth.md
Comment (p-bakker) [reply docs/sources/auth/generic-oauth.md]: done

File: docs/sources/auth/generic-oauth.md
Comment (p-bakker) [reply docs/sources/auth/generic-oauth.md]: done

File: docs/sources/auth/generic-oauth.md
Comment (p-bakker) [reply docs/sources/auth/generic-oauth.md]: done

**Review by achatterjee-grafana — APPROVED** [review]
LGTM

**Comment by achatterjee-grafana** [comment]
I see the checks are still failing. Did you merge the master to your local? Thanks.

**Comment by p-bakker** [comment]
Its still failing because of https://github.com/grafana/grafana/pull/32888#issuecomment-817906876

Was told to hold off on fixing that until a decision was made whether this functionality was desired or not.

Given that you approved it, I assume that's a yes now? If so, I'll go and fix the config type and make sure the build passes

**Comment by achatterjee-grafana** [comment]
Yes, once you do the merge from the master, I can take care of this PR. Thank you for your contribution!

**Comment by aknuds1** [comment]
> Its still failing because of [#32888 (comment)](https://github.com/grafana/grafana/pull/32888#issuecomment-817906876)
> 
> Was told to hold off on fixing that until a decision was made whether this functionality was desired or not.
> 
> Given that you approved it, I assume that's a yes now? If so, I'll go and fix the config type and make sure the build passes

@p-bakker @achatterjee-grafana it still needs signing off by a developer.

**Comment by p-bakker** [comment]
@aknuds1 I assume your comment means I should still hold off on trying to fix the settings thing that breaks the pipeline, correct?

I did look at it, but as for now I haven't figured out exactly yet how to add ` ` to `hs.Cfg`

**Comment by aknuds1** [comment]
@p-bakker yeah sorry, we're going to have to review and test this when time allows. Now we're busy prepping for v8. Would you like me to push a fix of the merge conflicts?

**Comment by p-bakker** [comment]
Yeah, if you could, I'd appreciate it!

**Comment by aknuds1** [comment]
I've gathered in the meantime that this might be a bit of a corner case, and we'd like to see more interest in the feature request at least before accepting the PR.

**Comment by marefr** [comment]
I added a https://github.com/grafana/grafana/issues/32887#issuecomment-829391304 in the referenced issue trying to get some better understanding of the feature request.

**Comment by marefr** [comment]
Thanks for your contribution. However, we haven't received any response to https://github.com/grafana/grafana/issues/32887#issuecomment-829391304 so I'll close this for now. Closing doesn't mean we couldn't re-open it, but suggest continue the discussions in referenced issue.

**Pull request closed without merging**
